solr/core/src/java/org/apache/solr/handler/component/UBIComponent.java

```diff
@@ -0,0 +1,424 @@
+  public void prepare(ResponseBuilder rb) throws IOException {
+    SolrParams params = rb.req.getParams();
+    if (!params.getBool(COMPONENT_NAME, false)) {
+      return;
+    }
+    rb.setNeedDocList(true);
+  }
+
+  @Override
+  public void process(ResponseBuilder rb) throws IOException {
+    SolrParams params = rb.req.getParams();
+    if (!params.getBool(COMPONENT_NAME, false)) {
+      return;
+    }
+
+    doStuff(rb);
+  }
+
+  @Override
+  public int distributedProcess(ResponseBuilder rb) throws IOException {
+
+    SolrParams params = rb.req.getParams();
+    if (!params.getBool(COMPONENT_NAME, false)) {
+      return ResponseBuilder.STAGE_DONE;
+    }
+
+    if (rb.stage != ResponseBuilder.STAGE_GET_FIELDS) {
+      return ResponseBuilder.STAGE_DONE;
+    }
+
+    doStuff(rb);
+
+    return ResponseBuilder.STAGE_DONE;
+  }
+
+  public void doStuff(ResponseBuilder rb) throws IOException {
+
+    // not sure why but sometimes we get it twoice...  how can a response have the
+    // the same component run twice?
+    if (rb.rsp.getValues().get("ubi") != null) {
+      return;
+    }
+    SolrParams params = rb.req.getParams();
+
+    SolrIndexSearcher searcher = rb.req.getSearcher();
+
+    String queryId = params.get(QUERY_ID);
+    UBIQuery ubiQuery = new UBIQuery(queryId);
+
+    ubiQuery.setUserQuery(params.get(USER_QUERY));
+    ubiQuery.setApplication(params.get(APPLICATION));
+    if (ubiQuery.getApplication() == null) {
+      ubiQuery.setApplication(
+          rb.isDistrib
+              ? rb.req.getCloudDescriptor().getCollectionName()
+              : searcher.getCore().getName());
+    }
+
+    String queryAttributes = params.get(QUERY_ATTRIBUTES);
+
+    if (queryAttributes != null && queryAttributes.toString().startsWith("{")) {
+      // Look up the original nested JSON format, typically passed in
+      // via the JSON formatted query.
+      @SuppressWarnings("rawtypes")
+      Map jsonProperties = rb.req.getJSON();
+      if (jsonProperties.containsKey("params")) {
+        @SuppressWarnings("rawtypes")
+        Map paramsProperties = (Map) jsonProperties.get("params");
+        if (paramsProperties.containsKey(QUERY_ATTRIBUTES)) {
+          @SuppressWarnings("rawtypes")
+          Map queryAttributesAsMap = (Map) paramsProperties.get(QUERY_ATTRIBUTES);
+          ubiQuery.setQueryAttributes(queryAttributesAsMap);
+        }
+      }
+    }
+
+    ResultContext rc = (ResultContext) rb.rsp.getResponse();
+    DocList docs = rc.getDocList();
+    // DocList docs = rb.getResults().docList;
+
+    String docIds = extractDocIds(docs, searcher);
+    ubiQuery.setDocIds(docIds);
+
+    addUserBehaviorInsightsToResponse(ubiQuery, rb);
+    recordQuery(ubiQuery);
+  }
+
+  private void recordQuery(UBIQuery ubiQuery) throws IOException {
+    TupleStream stream;
+
+    stream = constructStream(streamFactory, streamExpression);
+
+    streamContext.put("ubi-query", ubiQuery);
+    stream.setStreamContext(streamContext);
```

Comment:
I know nothing about streams, but is it possible to submit UbiQuery here, which isn't a subclass of any serializable framework? You know passing a ref might work locally, but not it remote/distributed env.

Reply:
GOod question.   I *believe* that the `getTuple` method will be immediately run, which means this code doesn't actually get run in a distributed sense..  I.e, the `ubiQuery`  object that we put into the `streamContext` is immediately read back out in `getTuple()` method..    That is the job of the `UBIQueryStream` class, to convert the `UBIQuery` found in the context into a `Tuple` that is used by streaming expressons.

I am going to try actually making `UBIQuery` and `UBIQueryStream` inner classes of `UBIComponent` to see how that looks...

Reply:
ok. got it. found `toMap()`. Shouldn't it yield docIds as well? I can't see refs to this field there https://github.com/epugh/solr/blob/99d6b7a7eb7b28a92f4cb36d4a525f8b901ba93c/solr/core/src/java/org/apache/solr/handler/component/UBIQuery.java#L103

Reply:
That's a fair question...   OMG...     I *can't* believe that I forgot to add the damn doc ids...   I suck.   

Thanks for spotting that!   How embarrassing!   yeah...

Reply:
And fixed.